src/packbuilder.rs

```diff
@@ -0,0 +1,358 @@
+    /// referenced trees and blobs.
+    pub fn insert_tree(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_tree(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// Insert a commit object. This will add a commit as well as the completed
+    /// referenced tree.
+    pub fn insert_commit(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_commit(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// Insert objects as given by the walk. Those commits and all objects they
+    /// reference will be inserted into the packbuilder.
+    pub fn insert_walk(&mut self, walk: &mut Revwalk) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_walk(self.raw, walk.raw()));
+        }
+        Ok(())
+    }
+
+    /// Recursively insert an object and its referenced objects. Insert the
+    /// object as well as any object it references.
+    pub fn insert_recursive(&mut self,
+                            id: Oid,
+                            name: Option<&str>)
+                            -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_recur(self.raw,
+                                                        &*id.raw(),
+                                                        name));
+        }
+        Ok(())
+    }
+
+    /// Write the contents of the packfile to an in-memory buffer. The contents
+    /// of the buffer will become a valid packfile, even though there will be
+    /// no attached index.
+    ///
+    /// `progress` will be called with progress information during pack
+    /// building. Be aware that this is called inline with pack building
+    /// operations, so performance may be affected.
+    pub fn write_buf(&mut self,
+                     buf: &mut Buf,
+                     progress: Option<&mut ProgressCb>)
+                     -> Result<(), Error> {
+        let has_progress = progress.is_some();
+        if let Some(mut progress) = progress {
+            let ptr = &mut progress as *mut _;
+            let progress_c = Some(progress_c as raw::git_packbuilder_progress);
+            unsafe {
+                try_call!(raw::git_packbuilder_set_callbacks(self.raw,
+                                                             progress_c,
+                                                             ptr as *mut _));
+            }
+        }
+        unsafe {
+            try_call!(raw::git_packbuilder_write_buf(buf.raw(), self.raw));
+        }
+        if has_progress {
+            unsafe {
+                try_call!(raw::git_packbuilder_set_callbacks(self.raw,
+                                                             None,
+                                                             ptr::null_mut()));
+            }
+        }
```

Comment:
Oh right, in the past the bounds I've used are:

```rust
pub fn foo<F>(&mut self, b: &'a mut F) where F: FnMut(...)
```

That way you call it as:

```rust
let mut cb = || { ... };
let obj = ...;
obj.foo(&mut cb);
```

The part about `'a` and `'b` is also kinda unfortunate, and I think there may not be a great way around that. For now, though, maybe we could remove this argument? We could always add another argument in another function in the future, or if perhaps we figure out how to pass the progress callback as a separate function.

Or is it basically necessary to expose the progress callback to land?

> Alternative would be to keep the current API but explicitly null out the callback if None is passed

I'd be somewhat worried as we'd still be storing dangling pointers, and some other function in the future may eventually start calling the progress callback.

I think instead a good solution here is to store the result of the underling raw call, unset the callbacks, and then return the original result (basically don't use `try!`